python/cudf/cudf/core/index.py

```diff
@@ -110,6 +110,8 @@ def _index_from_data(data: MutableMapping, name: Any = None):
             index_class_type = CategoricalIndex
         elif isinstance(values, IntervalColumn):
             index_class_type = IntervalIndex
+        else:
+            index_class_type = GenericIndex
```

Comment:
Reviewers: should we also try to fix the pivot problems I noted in #11148 before merging this, or is this PR a good scope? It removes the `UnboundLocalError` and it gives a result instead of raising, but there are some lingering issues with index/column values in `df.pivot`. I think we can punt on fixing those for a second PR but I would like another opinion.

Reply:
I agree, let's do pivot table fixes in a separate PR.

You have identified the right problem, but I'm not sure that this fix is correct. The problem we're running into here is the case where we have a column whose dtype doesn't map to a known index type, right? Specifically, there's no `ListIndex` or `StructIndex`? How does this case work in pandas? Does it create a GenericIndex of dtype object? If so, do we need to define new custom index types for our dtypes that pandas doesn't support? With your current patch, what works and what doesn't?

Reply:
Sorry I have been meaning to respond to this since you posted and it kept falling off my radar.

So there are definitely issues here caused by the lack of alignment. First and foremost is the fact that `cudf.Index` can't actually be created because its `__new__` always triggers the creation of a different index type. This also means that we differ from pandas in one or two other places, as noted in #11080. One other way that this manifests is with boolean indexes: 

```
>>> import pandas as pd
>>> pd.Index([True, False, True])
Index([True, False, True], dtype='object')
>>> import cudf
>>> cudf.Index([True, False, True])
GenericIndex([True, False, True], dtype='bool')
```

I think there are basically two questions associated with this PR:
1. The short term question: Are we OK with returning a GenericIndex with dtype `list` as a stopgap solution? I think the answer is yes. It's not perfect in terms of pandas alignment or in terms of exposing a class that I'd consider to be pretty internal, but it's better than failing for now. From that perspective, "what works and what doesn't" for me basically boils down to "does making a GenericIndex with dtype `list` work about as well as a GenericIndex with dtype `bool`". How does it behave if you use that index with a Series and try really basic operations like indexing with `__getitem__`/loc/iloc, or if you do a simple reduction operation? If those don't immediately cause errors, I think we can merge this PR for now.
2. The long term question: Do we need first-class support for indexes with dtypes that aren't supported by pandas? I think that given this issue and #11080 we probably need to think about this question soon. @shwina probably has some opinions here. I don't think resolving this question should be a blocker for this PR, though.

Reply:
> Are we OK with returning a GenericIndex with dtype list as a stopgap solution

Yes, I think that's fine.

>  Do we need first-class support for indexes with dtypes that aren't supported by pandas?

We already have this (`StringIndex`) and people generally don't seem to complain about the difference from Pandas. So I'd be in favor of `ListIndex`/`StructIndex` types.

Reply:
Cool, so @bdice I think your action item here is to just play with a Series/DataFrame that has a GenericIndex of dtype list and make sure it doesn't just immediately explode. As long as it's not something that just immediately borks all workflows I think we can merge this change and address the other index types later.

Reply:
Awesome! I'll follow up with a snippet of what I try. (And maybe add tests? Or wait until we "officially" support it with a new class?)

Reply:
I would recommend only testing cases where a user could fall backwards into creating this type of index (e.g. the pivot case). In practice that may be the only example we have right now, which is fine. IMO no need to explicitly test the functionality of the GenericIndex for now.